Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt

```diff
@@ -0,0 +1,117 @@
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.flow.firstOrNull
+import kotlinx.coroutines.withContext
+import org.joda.time.Duration
+import org.joda.time.Instant
+import timber.log.Timber
+import java.lang.reflect.Modifier.PRIVATE
+import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
+
+class CheckInWarningMatcher @Inject constructor(
+    private val checkInsRepository: CheckInRepository,
+    private val traceTimeIntervalWarningRepository: TraceTimeIntervalWarningRepository,
+    private val dispatcherProvider: DispatcherProvider
+) {
+    suspend fun execute(): List<CheckInWarningOverlap> {
+        val checkIns = checkInsRepository.allCheckIns.firstOrNull() ?: return emptyList()
+
+        val warningPackages = traceTimeIntervalWarningRepository.allWarningPackages.firstOrNull() ?: return emptyList()
+
+        val splitCheckIns = checkIns.flatMap { it.splitByMidnightUTC() }
+
+        return createMatchingLaunchers(splitCheckIns, warningPackages, dispatcherProvider.IO)
+            .awaitAll()
+            .flatten()
+    }
+}
+
+suspend fun createMatchingLaunchers(
+    checkIns: List<CheckIn>,
+    warningPackages: List<TraceTimeIntervalWarningPackage>,
+    coroutineContext: CoroutineContext
+): Collection<Deferred<List<CheckInWarningOverlap>>> {
+
+    val launcher: CoroutineScope.(
+        List<CheckIn>,
+        List<TraceTimeIntervalWarningPackage>
+    ) -> Deferred<List<CheckInWarningOverlap>> =
+        { list, packageChunk ->
+            async {
+                packageChunk.flatMap {
+                    findMatches(list, it)
+                }
+            }
+        }
+
+    // at most 4 parallel processes
+    val chunkSize = (checkIns.size / 4) + 1
+
+    return warningPackages.chunked(chunkSize).map { packageChunk ->
+        withContext(context = coroutineContext) {
+            launcher(checkIns, packageChunk)
+        }
+    }
+}
+
+@VisibleForTesting(otherwise = PRIVATE)
+internal suspend fun findMatches(
+    checkIns: List<CheckIn>,
+    warningPackage: TraceTimeIntervalWarningPackage
+): List<CheckInWarningOverlap> {
+    return warningPackage
+        .extractTraceTimeIntervalWarnings()
+        .flatMap { warning ->
+            checkIns
+                .mapNotNull { checkIn ->
+                    checkIn.calculateOverlap(warning).also { overlap ->
+                        if (overlap == null) {
+                            Timber.d("No match/overlap found for $checkIn and $warning")
+                        } else {
+                            Timber.i("Overlap found $overlap")
+                        }
+                    }
+                }
+        }
+}
+
+@VisibleForTesting(otherwise = PRIVATE)
+internal fun CheckIn.calculateOverlap(
+    warning: TraceWarning.TraceTimeIntervalWarning
+): CheckInWarningOverlap? {
+
+    if (warning.locationGuidHash != locationGuidHash) return null
```

Comment:
The techspec mentions to return `0`, not null. So is this a wanted behaviour here?

Reply:
https://github.com/corona-warn-app/cwa-app-tech-spec/blob/proposal/event-registration-mvp/docs/spec/event-registration-client.md#calculate-overlap-of-checkin-and-tracetimeintervalwarning

Reply:
Yes, nulls are skipped for further processing